Create v1 Collect workspace Bottom Up flow

mountiny opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

In this PR we are implementing a basic v1 bottom up flow in which users can:
- choose to pay p2p IOU report with business bank account
- collect workspace is created for them as admin and the submitter as employee
- the IOU report is converted to an expense report, the parent report changes to the new workspace chat, change the report to tal to negative amount as its done in case of expense reports
- the transaction data is updated to match expected transaction data for a workspace (amount is changed to negative value)
- We are adding a new Moved report action which is used to link the DM with the new workspace chat


### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/30702


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

For testing purposes as Expensify Engineer locally set the `skipGuidesAssignent` to true on [this line](https://github.com/Expensify/Web-Expensify/blob/4efe6a67862dd94f17341b72e02d369ce736e1c4/lib/PolicyAPI.php#L258) as most likely the guide assignment might not be working well for you locally and it could cause the call to fail.

0. Set your location/ VPN to the USA (thats to ensure the policy currency is set to USD which is the only currency we support for the business bank accounts in NewDot right now)
1. Create 2 accounts, one will be used as employee and second as an admin (now they are just a normal pals, friends, nothing more than that, but we will change them to an admin and employee in this flow)
2. Request 1:1 money as employee from the admin (the IOU report can only have "one-way" requests, if the admin would also request some money from the employee, this flow is not allowed/accessible)
3. As admin, go to the 1:1 chat and click `Pay With Expensify`
4. Choose to pay with Business Bank account
5. Verify you have been brought to the Bank account flow
6. Close the bank account flow
7. Verify you can see a new report message in the 1:1 chat which links you to the workspace chat of the new employee
8. Click on the link (when developing in localhost against staging api, the link will point to staging so you will have to update it)
9. Verify you can see a report preview for the same amount request in the expense report
10. Verify you can navigate to the expense report both as admin and employee
11. Verify the expense report and workspace chat correctly show the default workspace avatar in the LHN

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as tests

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
12. Upload an image via copy paste
13. Verify a modal appears displaying a preview of that image
--->

Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -

https://github.com/Expensify/App/assets/36083550/7b0c2633-c4e0-4a9e-ba50-3ff2d01a09a5

->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

mountiny committed:
Add a new button to the payment methods popover

mountiny committed:
Add the business bank account option

mountiny committed:
Work on the Bottom Up flow command

mountiny committed:
Merge main

mountiny committed:
Merge main

puneetlath commented on src/CONST.ts:
```diff
@@ -1113,6 +1113,7 @@ const CONST = {
     PAYMENT_METHODS: {
         DEBIT_CARD: 'debitCard',
         BANK_ACCOUNT: 'bankAccount',
```
Should we rename this to `BANK_ACCOUNT ` just for clarity?

puneetlath commented on src/components/AddPaymentMethodMenu.js:
```diff
@@ -61,12 +78,21 @@ function AddPaymentMethodMenu(props) {
                         props.onItemSelected(CONST.PAYMENT_METHODS.BANK_ACCOUNT);
                     },
                 },
+                ...(ReportUtils.isIOUReport(this.props.iouReport) && ReportActionsUtils.hasRequestFromPayer(lodashGet(this.props.iouReport, 'reportID', 0), this.props.session.accountID)
```
I'm a bit confused by `hasRequestFromPayer`. Wouldn't this be checking if the current user that is clicking the pay button as made the request?

mountiny replied:
should be negated yes

puneetlath commented on src/libs/actions/Policy.js:
```diff
@@ -1360,6 +1360,267 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
     return lodashUnion([category], policyRecentlyUsedCategories);
 }
 
+function startCollectBottomUpFlow(iouReport, firstName, lastName) {
```
I would call this something more specific to the action the user is taking. Some options: 
- payIOUWithBusinessBankAccount (I think this is most accurate for the action they are taking and everything else that happens is the side-effect of this action) 
- createWorkspaceFromIOUPayment
- createWorkspaceFromBBAPaymentOfIOU
-

mountiny replied:
using the `createWorkspaceFromIOUPayment`

puneetlath commented on src/libs/actions/Policy.js:
```diff
@@ -1360,6 +1360,267 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+    // Next we need to convert the IOU report to Expense report and clean up the DM chat
+    // Get the 1on1 chat where the request was originally made
+    const chatReportID = iouReport.chatReportID;
+    ReportUtils.getReport(iouReport.chatReportID);
```
```suggestion
    ReportUtils.getReport(chatReportID);
```

puneetlath commented on src/libs/actions/Policy.js:
```diff
@@ -1360,6 +1360,267 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+    ];
+
+    API.write(
+        'BottomUpCollectFlow',
```
I would similarly rename this to something like payIOUWithBusinessBankAccount

mountiny committed:
Fix merge conflict

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Fix issues in the flow

mountiny committed:
Update the personal bank account const name

mountiny committed:
Flow clean up

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Offer correct payment options

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Make sure to update the report total and transaction amounts

puneetlath commented on src/libs/ReportActionsUtils.ts:
```diff
@@ -616,6 +616,26 @@ function isNotifiableReportAction(reportAction: OnyxEntry<ReportAction>): boolea
+ * @param currentAccountID
+ * @returns
+ */
+function hasRequestFromPayer(reportID: string, currentAccountID: number): boolean {
```
This method doesn't actually check if the currentAccount is the payer, so I think this name is more accurate. 
```suggestion
function hasRequestFromCurrentAccount(reportID: string, currentAccountID: number): boolean {
```

puneetlath commented on src/libs/actions/Policy.js:
```diff
@@ -1360,6 +1360,280 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+        expenseCreatedReportActionID: workspaceChatCreatedReportActionID,
+    } = ReportUtils.buildOptimisticWorkspaceChats(policyID, workspaceName);
+
+    // Create the workspace chat for the employee
```
NAB just to make clearer that the current user isn't the employee
```suggestion
    // Create the workspace chat for the employee whose IOU is being paid
```

puneetlath reviewed (dismissed):
Looking good! Just some minor feedback.

mountiny committed:
Add moved report action

mountiny committed:
Fix crashes

mountiny commented on src/libs/actions/Policy.js:
```diff
@@ -1360,6 +1365,363 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+    );
+
+    // Navigate to the bank account set up flow for this specific policy
+    Navigation.navigate(ROUTES.BANK_ACCOUNT_WITH_STEP_TO_OPEN.getRoute('', policyID));
```
Sounds good, I agree!

mountiny committed:
Merge main

mountiny committed:
Move the navigation action out of Policy action

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Remove logs

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny committed:
Clean up the code

mountiny committed:
Fix lint and clean up some code

mountiny committed:
Fix lint

mountiny committed:
Merge branch 'main' into vit-bottomUp

mountiny commented on src/components/AddPaymentMethodMenu.js:
```diff
@@ -37,17 +44,27 @@ const propTypes = {
     anchorRef: refPropTypes,
 
+    /** Session info for the currently logged in user. */
+    session: PropTypes.shape({
```
Good point, I think that can be kept out of scope for this PR given its in a lot of places

mountiny commented on src/components/AddPaymentMethodMenu.js:
```diff
@@ -37,17 +44,27 @@ const propTypes = {
     betas: [],
     anchorRef: () => {},
+    session: {
+        accountID: 0,
```
Can be an empty object, just going to have to access the accountID safely where I call it, updating

mountiny committed:
Destructure props and change session propTypes in the AddPaymentMethodMenu

mountiny commented on src/libs/ReportUtils.js:
```diff
@@ -2667,6 +2669,55 @@ function buildOptimisticApprovedReportAction(amount, currency, expenseReportID)
+        originalMessage,
+        message: [
+            {
+                html: `moved the report to the <a href='${CONST.NEW_EXPENSIFY_URL}r/${newParentReportID}' target='_blank' rel='noreferrer noopener'>${policyName}</a> workspace`,
+                text: `moved the report to the ${policyName} workspace`,
```
For v1 they are fine to be in English, its a new action type to be handled in NewDot and we would make these polish adjustments in v2. We have been collecting the improvements in the tracking issue https://github.com/Expensify/App/issues/31314 which is added to the wave9 project for once this will be worked on

mountiny commented on src/libs/actions/Policy.js:
```diff
@@ -1458,6 +1461,375 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+    });
+
+    // The expense report transactions need to have the amount reversed to negative values
+    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReportID);
```
Would that be beneficial here? we do not need to have another connection to Transactions collection in Policy.js but just get all the transactions linked to the report, right?

tgolen replied:
There are several benefits to adding a new connection in this file:

- The local structure can be optimized for reportID lookup (ie. store them in a map like `{<reportID>: transactions}` rather than having two `filters()` run every time in `getAllReportTransactions()`
- The action files remain decoupled (something we do a terrible job of)
- Decoupled actions keep regressions to a minimum

mountiny commented on src/libs/actions/Policy.js:
```diff
@@ -1458,6 +1461,375 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+    const reportTransactions = TransactionUtils.getAllReportTransactions(iouReportID);
+    _.each(reportTransactions, (transaction) => {
+        optimisticData.push({
+            onyxMethod: Onyx.METHOD.MERGE,
```
Sounds good

mountiny commented on src/libs/actions/Policy.js:
```diff
@@ -1458,6 +1461,375 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+
+    const optimisticData = [
+        {
+            onyxMethod: Onyx.METHOD.SET,
```
I think it should, although I did take this from the CreateWorkspace so not sure if there were any problems with using merge, I will try

tgolen replied:
I actually had a GH at one point to go and fix a lot of that `CreateWorkspace` stuff to use merge, but I wasn't ever able to prioritize it. It is a bad pattern that shouldn't be followed due to the inconsistencies of working with both merge/set and many times `set()` is used as a workaround hack for things like wanting the operation to be sync instead of async, and not taking the time to make what is relying on the data being in Onyx to handle the async operation.

tgolen replied:
I think we should look at other solutions. The purpose of `SET` isn't to fix navigation issues.

mountiny committed:
Handle transactions in merge collection, prettier and translations

mountiny committed:
Clean up the GBR in the bottom up flow

mountiny committed:
Add props for onItemSelected

mountiny commented:
@tgolen Thank you for valuable review! I think I have addressed all your comments so feel free to take another look if you can.

@akinwale this is ready for you to test, you can point you api to staging just to be on the earliest version. there is one polish item which is already being addressed in the API and hat is the fact that the report preview does not get cleared from the 1:1 chat for the employee when the bottom up flow is executed. This should be fixed soon once pr is merged and deployed in the backend.

akinwale commented:
@mountiny I would need accounts set up with the admin / employee roles to be able to run the test steps, as I am unable to do this on a free policy (checking OldDot). I have previously used a free trial to set up a Control policy.

tgolen reviewed (dismissed):
This is a doozy of an optimistic action! Good job on it. Thanks for making those changes. I'll approve and leave the NABs to you.

mountiny committed:
Merge branch 'main' into vit-bottomUp

akinwale commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/4160319/8cfa8c44-ff5b-4189-a1d3-888efd15e97c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/4160319/40bb5051-c3d8-450e-9ec4-f475eb82a6e3

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/4160319/523024f1-b340-4b7a-a69f-2c00d27511db

https://github.com/Expensify/App/assets/4160319/668c8751-13bf-44c6-9dc3-0bb7c4988c0c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/4160319/486d518e-7931-467d-af65-aa25bd3826fb

</details>

akinwale reviewed (approved):
LGTM.

Tests well with no issues.

Julesssss commented on src/libs/ReportUtils.js:
```diff
@@ -2700,6 +2702,55 @@ function buildOptimisticApprovedReportAction(amount, currency, expenseReportID)
+        avatar: lodashGet(currentUserPersonalDetails, 'avatar', UserUtils.getDefaultAvatarURL(currentUserAccountID)),
+        isAttachment: false,
+        originalMessage,
+        message: [
```
Having a const message makes this slightly more readable, if so it would sit nicely beneath the `policyName` const.

Julesssss commented on src/components/KYCWall/BaseKYCWall.js:
```diff
@@ -93,10 +95,19 @@ class KYCWall extends React.Component {
         } else if (paymentMethod === CONST.PAYMENT_METHODS.DEBIT_CARD) {
             Navigation.navigate(this.props.addDebitCardRoute);
+        } else if (paymentMethod === CONST.PAYMENT_METHODS.BUSINESS_BANK_ACCOUNT) {
+            if (ReportUtils.isIOUReport(this.props.iouReport)) {
```
Is this additional check necessary? We also make the same check within the function. 

I think it is because of the navigation. And it's better to check within the function instead of outside. But I just want to double-check

Julesssss reviewed (dismissed):
This PR introduces quite a lot and I found it a bit tricky to keep context in mind when reviewing due to the size of this PR. 

__________________

> 7. Verify you can see a new report message in the 1:1 chat which links you to the workspace chat of the new employee

A workspace appears to have been created (for the admin), but I'm not seeing any reportComment. I see a regular 1:1 IOU and when clicked, I'm taken to the Expense report without seeing the workspace avatar: 

<img width="898" alt="Screenshot 2023-11-17 at 09 57 58" src="https://github.com/Expensify/App/assets/10736861/47ad061c-6e38-4a92-a370-cd21fd781986">


I also see an error on the admins avatar, but I can't follow the red-brick-road as nothing shows up in settings: 

<img width="898" alt="Screenshot 2023-11-17 at 09 58 07" src="https://github.com/Expensify/App/assets/10736861/6980e43e-c9bf-411f-b1a2-26cb924cd434">

Julesssss commented:
To clarify, the other user doesn't see the workspace chat, just the admin. I also made sure to pull changes and update Web-E, Web-S, & Auth.

mountiny commented:
> To clarify, the other user doesn't see the workspace chat, just the admin. I also made sure to pull changes and update Web-E, Web-S, & Auth.

Interested this worked in my testing/ video.

Thanks for testing

mountiny commented:
@Julesssss did you follow this step when testing locally:

> For testing purposes as Expensify Engineer locally set the `skipGuidesAssignent` to true on [this line](https://github.com/Expensify/Web-Expensify/blob/4efe6a67862dd94f17341b72e02d369ce736e1c4/lib/PolicyAPI.php#L258) as most likely the guide assignment might not be working well for you locally and it could cause the call to fail.

mountiny committed:
Merge branch 'main' into vit-bottomUp

Julesssss commented:
> skipGuidesAssignent

yeah, I'm 99% sure about that. Possibly I didn't save the file, but I definitely applied that code.

Julesssss commented:
> did you follow this step when testing locally:

yeah, I'm 99% sure about that. Possibly I didn't save the file, but I definitely applied that code.

mountiny committed:
Address PR review feedback

mountiny commented:
@Julesssss I could not repro nor @akinwale had this same issue, I have applied your feedback, could you please test once more?

Julesssss commented:
Just to double check, this git diff is okay for testing?

```
$guideAssignmentData = [];
```

Julesssss reviewed (commented):
Same as before, so I just want to check I'm following the steps correctly... So I am supposed to exit the bank flow without adding a card?:

> 4. Choose to pay with Business Bank account
> 5. Verify you have been brought to the Bank account flow
> 6. Close the bank account flow

These are both brand new users without chat history and this is before I click the 'Pay with Expensify' button:

![Screenshot 2023-11-20 at 14 28 19](https://github.com/Expensify/App/assets/10736861/5da7119d-fac6-4c8f-a3f0-634ee7a86a94)

And here's what I see when I do this by clicking outside the bank account setup window: 

![Screenshot 2023-11-20 at 14 29 58](https://github.com/Expensify/App/assets/10736861/42946d89-9622-463e-a916-1607e6febe24)

mountiny commented:
> Just to double check, this git diff is okay for testing?

Yeah this should be ok, I will give this another go

mountiny commented:
@Julesssss I have tested again and based on logs there are some unrelated issues with queueing some jobs in VM, its always helpful to check your logs and see whats causing the issue when you run into this during testing locally.

I have to comment out these [two lines too](https://github.com/Expensify/Web-Expensify/blob/e3143300101444cc356fd0263ef95e2c71422346/lib/PolicyStore.php#L60-L64), which queue jobs for account review and to eventually increase subscription but that is not related to this PR. There must be something off with this the error goes to bedrock-php.

For example I cannot even create a workspace locally through the normal menu if I dont comment these two lines out so I dont thin kthis is a blocker, attaching a video of the flow that I captured during testing now:

https://github.com/Expensify/App/assets/36083550/1e235bc8-9815-4357-8cb5-91c33d2669e9

Julesssss reviewed (commented):
I tried again with those additional lines commented out and now the Admins chat is updated. However, the user still doesn't see the message until I sign out and back in. 

I see two errors, adding both but it looks like the 'Account not found' is the true cause after a brief look. 

I'll leave it up to you to decide on the merge, as you don't seem to have this issue.

<img width="1512" alt="Screenshot 2023-11-21 at 09 52 55" src="https://github.com/Expensify/App/assets/10736861/5b39de91-f861-4d08-b81c-d2b11913e6fc">
<img width="1512" alt="Screenshot 2023-11-21 at 09 53 31" src="https://github.com/Expensify/App/assets/10736861/e4f5eca3-eb73-406f-bc4d-424f05838664">

mountiny commented:
Hmm not sure this worked for me fine as seen in the video @akinwale could you please give this one more go and share if things work fine with staging API? Thank you!

mountiny commented:
@Julesssss fwiw the logs suggest you did not change to `$guideAssignmentData = [];` 🤔 as it fails assigning a guide

Julesssss commented:
> fwiw the logs suggest you did not change

I 100% did this. I was very careful to save the file given that this was the 3rd test

akinwale commented:
@mountiny Test results with staging server enabled. Video showing fresh accounts (left: admin, right: employee) included.

**Requesting (employee) user**
* Added to #announce room for the created workspace
* Collab. IOU report created (collaboration between user and workspace) 
* No workspace chat

**Requestee (admin) user**
* Added to #admin and #announce rooms for the workspace
* Collab. IOU report created
* IOU report moved
* No workspace chat

https://github.com/Expensify/App/assets/4160319/7cf1644e-c514-4dc7-94e3-61fbcc1b0871

puneetlath commented:
I see the employee's workspace chat there for both. Isn't it these? Seems like it's working correctly to me. 

<img width="1509" alt="Screenshot 2023-11-22 at 2 59 35 PM" src="https://github.com/Expensify/App/assets/1371996/ae9ae6c1-b6fd-41e5-a8b8-f2f522d67105">

puneetlath commented:
Oh, I guess the admin's own workspace chat doesn't seem to have been added. But the employee's does for both.

akinwale commented:
> Oh, I guess the admin's own workspace chat doesn't seem to have been added. But the employee's does for both.

Yeah, I assumed we're looking for the workspace's chat room which is different from those. I opened all the reports in the video just to make sure.

mountiny commented:
It was added optimistically but then disappeared, I think this might be something to do with the Collect policy settings. This would require a change in the backend I have created a follow up to investigate so we do not forget about this https://github.com/Expensify/App/issues/31754

Both users have workspace chat and seems like everything worked so I am going to go ahead with merge. Thanks everyone fore testing and reviewing 🚀

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6972419696) to staging by https://github.com/mountiny in version: 1.4.3-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|cancelled 🔪
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6972419696) to staging by https://github.com/mountiny in version: 1.4.3-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

situchan commented on src/components/AddPaymentMethodMenu.js:
```diff
@@ -37,42 +47,63 @@ const propTypes = {
+                          {
+                              text: translate('common.businessBankAccount'),
+                              icon: Expensicons.Building,
+                              onSelected: () => onItemSelected(CONST.PAYMENT_METHODS.BUSINESS_BANK_ACCOUNT),
```
This caused crash - #31797 as it fallback to throw here:
https://github.com/Expensify/App/blob/7a68a3d1003130e63065bcbef253c912a3431c08/src/pages/settings/Wallet/WalletPage/WalletPage.js#L210

luacmartins commented on src/components/KYCWall/BaseKYCWall.js:
```diff
@@ -93,10 +95,19 @@ class KYCWall extends React.Component {
+                Navigation.navigate(ROUTES.BANK_ACCOUNT_WITH_STEP_TO_OPEN.getRoute('', policyID));
+                return;
+            }
+            Navigation.navigate(this.props.addBankAccountRoute);
```
I think this default might be incorrect. I think this defaults to a personal bank account since we pass it [here](https://github.com/Expensify/App/blob/7a68a3d1003130e63065bcbef253c912a3431c08/src/pages/settings/Wallet/WalletPage/WalletPage.js#L381)

mountiny replied:
I think this should be correct because we pass it to the SettlementButton in both cases using this method which checks if the parent report is workspace chat or DM https://github.com/Expensify/App/blob/e909316f451a08aedd420c31f19cc3ea3cac0243/src/components/MoneyReportHeader.js#L94

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7009509621) to production by https://github.com/AndrewGable in version: 1.4.3-11 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

situchan commented on src/components/SettlementButton.js:
```diff
@@ -164,42 +159,19 @@ function SettlementButton({
-        if (isExpenseReport) {
-            buttonOptions.push(paymentMethods[CONST.IOU.PAYMENT_TYPE.VBBA]);
-        }
+        buttonOptions.push(paymentMethods[CONST.IOU.PAYMENT_TYPE.VBBA]);
```
This change caused [duplicated "Pay with Expensify" option](https://github.com/Expensify/App/issues/31829)

mountiny commented on src/libs/actions/Policy.js:
```diff
@@ -1458,6 +1463,403 @@ function buildOptimisticPolicyRecentlyUsedCategories(policyID, category) {
+
+    const optimisticData = [
+        {
+            onyxMethod: Onyx.METHOD.MERGE,
```
Thanks!

tgolen replied:
😢 This is multiple times I've seen this used and I think it is a hack and that we should really try to find some other real solution to the problem.

mountiny replied:
I did raise that concern in the linked issue, however, no other viable solution has been found in couple of weeks. Then given we already use the `SET` for classic policy creation (so not part of the bottom up flow), I agree to proceed with using `SET`.

tgolen replied:
Would you be OK with creating a GH issue with a solid problem statement and start getting some proposals for fixing it? I'd like to not just let this one fade away with others continuing to copy the same hack. I have some ideas like:

- If there was just a `merge()` then the data should be immediately (and synchronously) available in the Onyx cache. Let's find a way to use it faster
- Is the problem because subscribers aren't notified until the next-tick? If so, maybe we can add an option to Onyx to bypass for some scenarios (since it was only added as an optimization)

mountiny replied:
I have this saved, but I still did not get around to do this focusing on Ideal nav and waves in general. Noting previous comment from [another PR](https://github.com/Expensify/App/pull/30582#discussion_r1395850276) for future.
